### DISCUSSION TITLE
Leveraging paginated data in "item" page

### DISCUSSION THREAD

**Discussion by petersendidit** [TanStack/query#785] [General] [answered]
In the [Star Wars](https://react-query.tanstack.com/docs/examples/star-wars) example the code queries the list of films when clicking on the Films page. Then when selecting a individual film the code doesn't use the existing data about the film from the list, it has to wait for query that is specific for that single film before the page is shown. 

How do you see best leveraging the data already loaded from the list to "pre-populate" the data for the single file page? 
Should it pre-populate the queryCache in some way?

**Comment by fnn** [comment]
Use `initalData` for the item query and use the query data from the list to return the initial data.
``` javascript
initialData: () => queryCache.getQueryData("list").find((item) => item.id ===id)
```

I came across the problem with pagination because each page has its own query key.
In that case maybe the best way is to use `setQueryData` to save all items in one query key. 🤷

**Reply by fnn** [reply]
Because my filters (page, ...) are outside the list component, I just pass the filters to the item component to use `getQuerydata`.

Rough example:
``` javascript
const Parent: React.FC = () => {
  const [filters, setFilters] = React.useState("?limit=50");

  const handlePageChange = () => {
    setFilters(...);
  };

  return (
      <div>
        <div>filter component here</div>

        <List filters={filters} onPageChange={handlePageChange} />

        <Routes>
          <Route
            path=":id"
            children={<Item listQueryKey={filters} />} // Now the query inside the Item component can use correct queryKey
          />
        </Routes>
      </div>
  );
};
```

**Comment by petersendidit** [comment]
Seems like another option is to use `queryCache.setQueryData` on click in the place that has the data.  In the Star Wars example that would be changing this code
https://github.com/tannerlinsley/react-query/blob/270fa0f4233fba6b4f7d0c48d2e5e21ca82be7df/examples/star-wars/src/Films.js#L27
To this:
```jsx
<Link
  component={RouterLink}
  to={`/films/${filmId}`}
  onClick={() => { queryCache.setQueryData(`film-${filmId}`, film); }}
>
```

**Reply by fnn** [reply]
I really like that.

**Comment by johnslay** [comment, marked as answer]
How about this?
```
React.useEffect(() => {
  data.forEach(film => {
    queryCache.setQueryData(`film-${film.id}`, film)
  })
}, [data])
```

**Reply by tannerlinsley** [reply]
This is one of the patterns I suggest in the upcoming React Query Essentials course. It's even better if you put it in the core fetcher function, that way it's independent of the hook instance.
